bap-vibes/src/patch_c.ml

```diff
@@ -0,0 +1,1501 @@
+    SEQUENCE (simplify_casts_stmt s1, simplify_casts_stmt s2)
+  | IF (cond, st, sf) ->
+    IF (simplify_casts_exp cond,
+        simplify_casts_stmt st,
+        simplify_casts_stmt sf)
+  | GOTO _ as s -> s
+
+(* Find which vars are used *)
+
+module Used = struct
+
+  module Env = struct
+
+    type t = String.Set.t
+
+    let use (v : string) (env : t) : t = Set.add env v
+
+  end
+
+  include Monad.State.T1(Env)(Monad.Ident)
+  include Monad.State.Make(Env)(Monad.Ident)
+
+end
+
+type 'a used = 'a Used.t
+
+open Used.Let
+
+let rec used_exp : exp -> unit used = function
+  | UNARY (_, e, _) -> used_exp e
+  | BINARY (_, l, r, _) ->
+    let* () = used_exp l in
+    used_exp r
+  | CAST (_, e) -> used_exp e
+  | CONST_INT _ -> Used.return ()
+  | VARIABLE (v, _) -> Used.(update @@ Env.use @@ Theory.Var.name v)
+
+and used_stmt : stmt -> unit used = function
+  | NOP -> Used.return ()
+  | BLOCK (_, s) -> used_stmt s
+  | ASSIGN (_, e) -> used_exp e
+  | CALL (f, args) | CALLASSIGN (_, f, args) ->
+    let* () = used_exp f in
+    Used.List.iter args ~f:used_exp
+  | STORE (l, r) ->
+    let* () = used_exp l in
+    used_exp r
+  | SEQUENCE (s1, s2) ->
+    let* () = used_stmt s1 in
+    used_stmt s2
+  | IF (cond, st, sf) ->
+    let* () = used_exp cond in
+    let* () = used_stmt st in
+    used_stmt sf
+  | GOTO _ -> Used.return ()
+
+let rec remove_unused_temps (used : String.Set.t) : stmt -> stmt = function
+  | NOP -> NOP
+  | BLOCK (tenv, s) -> BLOCK (tenv, remove_unused_temps used s)
+  | ASSIGN ((v, _), _)
+    when Theory.Var.is_virtual v
+      && not (Set.mem used @@ Theory.Var.name v) -> NOP
+  | ASSIGN _ as s -> s
+  | (CALL _ | CALLASSIGN _) as s -> s
+  | STORE _ as s -> s
+  | SEQUENCE (s1, s2) ->
+    SEQUENCE (remove_unused_temps used s1, remove_unused_temps used s2)
+  | IF (cond, st, sf) ->
+    IF (cond, remove_unused_temps used st, remove_unused_temps used sf)
+  | GOTO _ as s -> s
+
+(* Translate a definition. *)
+let translate (patch : Cabs.definition) ~(target : Theory.target) : t KB.t =
+  let open KB.Let in
+  let* body = match patch with
+    | FUNDEF (_, b) -> KB.return b
+    | _ ->
+      let s = Utils.print_c Cprint.print_def patch in
+      Err.fail @@ Core_c_error (
+        sprintf "Patch_c.translate: unexpected patch shape:\n\n%s\n\n\
+                 expected a single function definition" s) in
+  (* Perform type-checking and elaboration. *)
+  let* (tenv, s), _ =
+    Transl.(Env.create ~target () |> run (translate_body body)) in
+  (* Perform some simplification passes. *)
+  let s = simplify_casts_stmt s in
+  let used = Monad.State.exec (used_stmt s) String.Set.empty in
+  let s = remove_unused_temps used s in
+  let s = simplify_nops s in
+  (* Success! *)
+  let prog = tenv, s in
+  Events.send @@ Rule;
+  Events.send @@ Info "Translated to the following PatchC program:";
+  Events.send @@ Info "";
+  Events.send @@ Info (to_string prog);
+  Events.send @@ Rule;
+  KB.return prog
```

Comment:
This unit, even in its infancy is huge, how large it will become in a few months when the PatcC language will grow with new features is hard to imagine, but I would imagine that it will take thousands of lines of code. It is better to split it into modules sooner than later.

Reply:
The main compiler pass has a large number of mutually recursive functions. I'm wondering if it makes sense (or is possible) to split them into modules. I know that recursive modules are a thing, but I wonder if they actually work that well or not.

Reply:
Made a ticket for this.